src/transport/raw/TCP.cpp

```diff
@@ -349,16 +351,45 @@ CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const Pe
             buffer->ConsumeHead(messageSize);
             SuccessOrExit(err);
 
-            err = endPoint->AckReceive(messageSize);
+            err = endPoint->AckReceive(payloadSize);
             SuccessOrExit(err);
             continue;
         }
 
-        // Buffer is incomplete if we reach this point
+        // If we reach this point, the head buffer does not contain a complete message.
+        // We may or may not have a complete message in a buffer chain.
         if (buffer->HasChainedBuffer())
         {
-            buffer->CompactHead();
-            continue;
+            if (payloadSize > System::PacketBuffer::kMaxSizeWithoutReserve)
+            {
+                // This message will not fit in a maximum-size buffer.
+                if (payloadSize <= buffer->TotalLength())
+                {
+                    // We have the oversize message in the current buffer chain.
+                    // Since it's oversize, it is definitely not a valid CHIP message, so we angrily toss it aside.
+                    buffer.Consume(payloadSize);
+                    err = endPoint->AckReceive(payloadSize);
```

Comment:
This logic path is not going to work.  `ContainsCompleteMessage` already evaluated that we do not have a complete message.  Thus we have received fewer than `payloadSize` bytes.  We shouldn't `buffer.Consume` more bytes than we have and we definitely cannot TCP ack more bytes than we've received.

In order to discard messages mid-payload, we'd need some sort of state machine to handle this.  Right now, all of our state is stored inside the buffer itself.

Reply:
Yes, I think we need state across calls. That should simplify getting the Acks right as well.